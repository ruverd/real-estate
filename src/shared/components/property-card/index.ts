export * from "./property-card";
export * from "./property-card-skeleton";

