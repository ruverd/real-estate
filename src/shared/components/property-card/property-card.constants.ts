import { Bath, BedDouble, Car, Ruler } from "lucide-react";

export const PROPERTY_FEATURES = [
  {
    icon: BedDouble,
    propertyKey: "bedrooms",
    label: "Bedrooms",
  },
  {
    icon: Bath,
    propertyKey: "bathrooms",
    label: "Bathrooms",
  },
  {
    icon: Ruler,
    propertyKey: "sqft",
    label: "Total area",
  },
  {
    icon: Car,
    propertyKey: "parking",
    label: "Parking",
  },
] as const;