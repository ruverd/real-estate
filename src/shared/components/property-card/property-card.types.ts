import type { Property } from "@/shared/types/property.types";
import type { LucideIcon } from "lucide-react";

export type PropertyFeatureOptions = 'bedrooms' | 'bathrooms' | 'sqft' | 'parking';

export type PropertyFeature = {
  icon: LucideIcon;
  propertyKey: PropertyFeatureOptions;
  label: string;
}

export interface PropertyFeatureProps extends Omit<PropertyFeature, 'propertyKey'> {
  value: string | number;
}

export interface PropertyCardProps {
  property: Property;
  showFeatures?: boolean;
  className?: string;
}