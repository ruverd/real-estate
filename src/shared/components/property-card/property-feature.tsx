import type { PropertyFeatureProps } from "./property-card.types";

export const PropertyFeature = ({ icon: Icon, value, label }: PropertyFeatureProps) => (
  <div className="flex flex-col items-center gap-1">
    <div className="flex items-center justify-center gap-1">
      <Icon className="size-5" />
      <span className="text-sm font-medium">{value}</span>
    </div>
    <span className="text-xs font-medium">{label}</span>
  </div>
);