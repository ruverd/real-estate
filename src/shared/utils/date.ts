import { differenceInDays, isValid, parseISO } from 'date-fns';

export const DAYS_THRESHOLD = 30;

export const isPropertyNew = (listedDate: string): boolean => {
  try {
    const parsedListedDate = parseISO(listedDate);

    console.log('parsedListedDate', parsedListedDate);
    console.log('isValid', isValid(parsedListedDate));
    
    if (!isValid(parsedListedDate)) {
      return false;
    }
    
    const daysDifference = differenceInDays(new Date(), parsedListedDate);

    console.log('daysDifference', daysDifference);
    
    return daysDifference <= DAYS_THRESHOLD;
  } catch {
    return false;
  }
};