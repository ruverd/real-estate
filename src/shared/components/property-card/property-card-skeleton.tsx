import { PROPERTY_FEATURES } from "../property-card/property-card.constants"
import { Card, CardContent, CardFooter, CardHeader } from "../ui/card"
import { Separator } from "../ui/separator"
import { Skeleton } from "../ui/skeleton"

export const PropertyCardSkeleton = () => {
  return (
    <Card className="overflow-hidden">
      <CardHeader className="relative p-0">
        <Skeleton className="h-[200px] w-full rounded-tl-xl rounded-tr-xl rounded-b-none" />
      </CardHeader>
    
      <CardContent className="p-4 flex flex-col gap-2">
        <div className="flex flex-col gap-2">
          <Skeleton className="h-[20px] w-full" />
          <Skeleton className="h-[20px] w-full" />
        </div>
        <Skeleton className="h-[24px] w-[100px]" />
        <Separator className="mt-2" />
      </CardContent>
      <CardFooter className="px-4 pt-0 pb-4 grid grid-cols-4 gap-2 text-sm text-foreground/80 w-full">
        {PROPERTY_FEATURES.map((feature) => (
          <Skeleton key={feature.propertyKey} className="h-[42px] w-full" />
        ))}
      </CardFooter>
    </Card>
  )
}