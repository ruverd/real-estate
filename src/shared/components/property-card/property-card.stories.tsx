import { MOCK_PROPERTIES } from "@/shared/mocks/property.mock";
import type { Meta, StoryObj } from "@storybook/react";
import { BrowserRouter } from "react-router-dom";

import { PropertyCard } from "./property-card";

const meta: Meta<typeof PropertyCard> = {
  title: "Components/PropertyCard",
  component: PropertyCard,
  tags: ["autodocs"],
  decorators: [
    (Story) => (
      <BrowserRouter>
        <div className="max-w-sm">
          <Story />
        </div>
      </BrowserRouter>
    ),
  ],
};

export default meta;
type Story = StoryObj<typeof PropertyCard>;

export const Default: Story = {
  args: {
    property: MOCK_PROPERTIES[0],
  },
};

export const WithoutFeatures: Story = {
  args: {
    property: MOCK_PROPERTIES[0],
    showFeatures: false,
  },
};

export const WithImageError: Story = {
  args: {
    property: {
      ...MOCK_PROPERTIES[0],
      thumbnailUrl: "https://invalid-image-url.jpg",
    },
  },
};
