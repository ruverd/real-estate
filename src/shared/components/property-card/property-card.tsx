import { useMemo, useState } from "react";
import { useNavigate } from "react-router-dom";

import { Badge } from "@/shared/components/ui/badge";
import { Card, CardContent, CardFooter, CardHeader } from "@/shared/components/ui/card";
import { isPropertyNew } from "@/shared/utils/date";
import { formatPrice } from "@/shared/utils/number";

import { Separator } from "../ui/separator";
import { PROPERTY_FEATURES } from "./property-card.constants";
import type { PropertyCardProps } from "./property-card.types";
import { PropertyFeature } from "./property-feature";

export function PropertyCard({ property, showFeatures = true }: PropertyCardProps) {
  const [imageLoaded, setImageLoaded] = useState(false);
  const [imageError, setImageError] = useState(false);
  const navigate = useNavigate();
  
  const renderNewTag = useMemo(() => {
    const showNewTag = isPropertyNew(property.dateListed);
    
    if (showNewTag) {
      return <Badge variant="secondary" className="absolute top-4 left-4">New</Badge>;
    }
    
    return null;
  }, [property.dateListed]);
  
  const formattedPrice = useMemo(() => formatPrice(property.salePrice), [property.salePrice]);
  
  const handleCardClick = () => {
    navigate(`/property/${property.id}`);
  };
  
  const handleKeyDown = (e: React.KeyboardEvent) => {
    if (e.key === "Enter" || e.key === " ") {
      e.preventDefault();
      handleCardClick();
    }
  };
  
  return (
    <Card 
      className="overflow-hidden cursor-pointer hover:-translate-y-2 transition-all duration-300"
      tabIndex={0}
      role="article"
      aria-label={`${property.title} in ${property.location} for ${formattedPrice}`}
      onClick={handleCardClick}
      onKeyDown={handleKeyDown}
    >
      <CardHeader className="relative p-0">
        {!imageLoaded && !imageError && (
          <div className="w-full h-[200px] bg-muted animate-pulse" data-testid="loading-placeholder" />
        )}
        {imageError && (
          <div className="w-full h-[200px] bg-muted flex items-center justify-center">
            <span className="text-muted-foreground" data-testid="image-error-message">Image not available</span>
          </div>
        )}
        <img 
          src={property.thumbnailUrl} 
          alt={property.title}
          className={`w-full h-[200px] object-cover ${!imageLoaded ? 'hidden' : ''}`}
          onLoad={() => setImageLoaded(true)}
          onError={() => setImageError(true)}
        />
        {renderNewTag}
      </CardHeader>

      <CardContent className="p-4 flex flex-col gap-2">
        <div className="flex flex-col">
          <h3 className="text-lg font-semibold text-foreground truncate">{property.title}</h3>
          <p className="text-sm font-display text-foreground/80 truncate">{property.location}</p>
        </div>
        <span className="text-xl font-display font-semibold">{formattedPrice}</span>
        {showFeatures && <Separator className="mt-2" />}
      </CardContent>
      
      {showFeatures && (
        <CardFooter className="px-4 pt-0 pb-4 grid grid-cols-4 gap-2 text-sm text-foreground/80 w-full">
          {PROPERTY_FEATURES.map((feature) => (
            <PropertyFeature
              key={feature.propertyKey}
              icon={feature.icon}
              value={property[feature.propertyKey]}
              label={feature.label}
            />
          ))}
        </CardFooter>
      )}
    </Card>
  );
}