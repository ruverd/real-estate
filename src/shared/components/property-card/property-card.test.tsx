import { fireEvent, render, screen, within } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { MOCK_PROPERTIES } from '@/shared/mocks/property.mock';
import { isPropertyNew } from '@/shared/utils/date';
import { formatPrice } from '@/shared/utils/number';

import { PropertyCard } from './property-card';
import { PROPERTY_FEATURES } from './property-card.constants';
import type { PropertyCardProps } from './property-card.types';

vi.mock('react-router-dom', () => ({
  useNavigate: () => vi.fn(),
}));

vi.mock('@/shared/utils/number', () => ({
  formatPrice: vi.fn().mockImplementation((price) => `$${price.toLocaleString()}`),
}));

vi.mock('@/shared/utils/date', () => ({
  isPropertyNew: vi.fn(),
}));

const mockProperty = MOCK_PROPERTIES[0];

const renderPropertyCard = (props?: PropertyCardProps) => {
  const defaultProps: PropertyCardProps = {
    property: mockProperty,
    showFeatures: true,
  };

  return render(<PropertyCard {...defaultProps} {...props} />);
};

describe('PropertyCard', () => {
  
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('should render property information correctly', () => {
    renderPropertyCard();
    
    expect(screen.getByText(mockProperty.title)).toBeInTheDocument();
    expect(screen.getByText(mockProperty.location)).toBeInTheDocument();
    expect(screen.getByText(formatPrice(mockProperty.salePrice))).toBeInTheDocument();
    
    const bedroomsSection = screen.getByText('Bedrooms').closest('div');
    const bathroomsSection = screen.getByText('Bathrooms').closest('div');
    const areaSection = screen.getByText('Total area').closest('div');
    const parkingSection = screen.getByText('Parking').closest('div');
    
    expect(within(bedroomsSection!).getByText(mockProperty.bedrooms.toString())).toBeInTheDocument();
    expect(within(bathroomsSection!).getByText(mockProperty.bathrooms.toString())).toBeInTheDocument();
    expect(within(areaSection!).getByText(mockProperty.sqft.toString())).toBeInTheDocument();
    expect(within(parkingSection!).getByText(mockProperty.parking.toString())).toBeInTheDocument();
    
    PROPERTY_FEATURES.forEach(feature => {
      expect(screen.getByText(feature.label)).toBeInTheDocument();
    });
  });

  it('should display the badge when property is new', () => {
    vi.mocked(isPropertyNew).mockReturnValue(true);
    
    renderPropertyCard();
    
    expect(screen.getByText('New')).toBeInTheDocument();
  });

  it('should not display the badge when property is not new', () => {
    vi.mocked(isPropertyNew).mockReturnValue(false);
    
    renderPropertyCard();
    
    expect(screen.queryByText('New')).not.toBeInTheDocument();
  });

  it('should not render features when showFeatures is false', () => {
    renderPropertyCard({ property: mockProperty, showFeatures: false });
    
    PROPERTY_FEATURES.forEach(feature => {
      expect(screen.queryByText(feature.label)).not.toBeInTheDocument();
    });
  });

  it('should have proper accessibility attributes', () => {
    renderPropertyCard();
    
    const card = screen.getByRole('article');

    expect(card).toHaveAttribute('tabIndex', '0');
    expect(card).toHaveAttribute('aria-label', 
      `${mockProperty.title} in ${mockProperty.location} for ${formatPrice(mockProperty.salePrice)}`
    );
  });

  it('should show loading state before image loads', () => {
    renderPropertyCard();
    
    const loadingPlaceholder = screen.getByTestId('loading-placeholder'); 
    const image = screen.getByAltText(mockProperty.title);

    expect(loadingPlaceholder).toBeInTheDocument();
    expect(image).toHaveClass('hidden');
  });

  it('should show error state when image fails to load', () => {
    renderPropertyCard();
    
    const image = screen.getByAltText(mockProperty.title);
    fireEvent.error(image);
    
    expect(screen.getByText('Image not available')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { asFragment } = renderPropertyCard();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should not render features when showFeatures is false', () => {
    renderPropertyCard({ property: mockProperty, showFeatures: false });

    expect(screen.queryByText('Features')).not.toBeInTheDocument();
  });
});
