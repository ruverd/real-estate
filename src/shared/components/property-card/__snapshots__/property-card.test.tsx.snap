// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PropertyCard > should match snapshot 1`] = `
<DocumentFragment>
  <div
    aria-label="Lovely House with Mountain View in 42 Rue de Rivoli, Paris, France for $349,000"
    class="rounded-xl border bg-card text-card-foreground shadow overflow-hidden cursor-pointer hover:-translate-y-2 transition-all duration-300"
    role="article"
    tabindex="0"
  >
    <div
      class="flex flex-col space-y-1.5 relative p-0"
    >
      <div
        class="w-full h-[200px] bg-muted animate-pulse"
        data-testid="loading-placeholder"
      />
      <img
        alt="Lovely House with Mountain View"
        class="w-full h-[200px] object-cover hidden"
        src="https://images.unsplash.com/photo-1588880331179-bc9b93a8cb5e?q=90&fm=jpg&w=300&fit=max"
      />
    </div>
    <div
      class="p-4 flex flex-col gap-2"
    >
      <div
        class="flex flex-col"
      >
        <h3
          class="text-lg font-semibold text-foreground truncate"
        >
          Lovely House with Mountain View
        </h3>
        <p
          class="text-sm font-display text-foreground/80 truncate"
        >
          42 Rue de Rivoli, Paris, France
        </p>
      </div>
      <span
        class="text-xl font-display font-semibold"
      >
        $349,000
      </span>
      <div
        class="bg-border shrink-0 data-[orientation=horizontal]:h-px data-[orientation=horizontal]:w-full data-[orientation=vertical]:h-full data-[orientation=vertical]:w-px mt-2"
        data-orientation="horizontal"
        data-slot="separator-root"
        role="none"
      />
    </div>
    <div
      class="items-center p-6 px-4 pt-0 pb-4 grid grid-cols-4 gap-2 text-sm text-foreground/80 w-full"
    >
      <div
        class="flex flex-col items-center gap-1"
      >
        <div
          class="flex items-center justify-center gap-1"
        >
          <svg
            class="lucide lucide-bed-double size-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2 20v-8a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v8"
            />
            <path
              d="M4 10V6a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v4"
            />
            <path
              d="M12 4v6"
            />
            <path
              d="M2 18h20"
            />
          </svg>
          <span
            class="text-sm font-medium"
          >
            4
          </span>
        </div>
        <span
          class="text-xs font-medium"
        >
          Bedrooms
        </span>
      </div>
      <div
        class="flex flex-col items-center gap-1"
      >
        <div
          class="flex items-center justify-center gap-1"
        >
          <svg
            class="lucide lucide-bath size-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 4 8 6"
            />
            <path
              d="M17 19v2"
            />
            <path
              d="M2 12h20"
            />
            <path
              d="M7 19v2"
            />
            <path
              d="M9 5 7.621 3.621A2.121 2.121 0 0 0 4 5v12a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2v-5"
            />
          </svg>
          <span
            class="text-sm font-medium"
          >
            3
          </span>
        </div>
        <span
          class="text-xs font-medium"
        >
          Bathrooms
        </span>
      </div>
      <div
        class="flex flex-col items-center gap-1"
      >
        <div
          class="flex items-center justify-center gap-1"
        >
          <svg
            class="lucide lucide-ruler size-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21.3 15.3a2.4 2.4 0 0 1 0 3.4l-2.6 2.6a2.4 2.4 0 0 1-3.4 0L2.7 8.7a2.41 2.41 0 0 1 0-3.4l2.6-2.6a2.41 2.41 0 0 1 3.4 0Z"
            />
            <path
              d="m14.5 12.5 2-2"
            />
            <path
              d="m11.5 9.5 2-2"
            />
            <path
              d="m8.5 6.5 2-2"
            />
            <path
              d="m17.5 15.5 2-2"
            />
          </svg>
          <span
            class="text-sm font-medium"
          >
            4638
          </span>
        </div>
        <span
          class="text-xs font-medium"
        >
          Total area
        </span>
      </div>
      <div
        class="flex flex-col items-center gap-1"
      >
        <div
          class="flex items-center justify-center gap-1"
        >
          <svg
            class="lucide lucide-car size-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 17h2c.6 0 1-.4 1-1v-3c0-.9-.7-1.7-1.5-1.9C18.7 10.6 16 10 16 10s-1.3-1.4-2.2-2.3c-.5-.4-1.1-.7-1.8-.7H5c-.6 0-1.1.4-1.4.9l-1.4 2.9A3.7 3.7 0 0 0 2 12v4c0 .6.4 1 1 1h2"
            />
            <circle
              cx="7"
              cy="17"
              r="2"
            />
            <path
              d="M9 17h6"
            />
            <circle
              cx="17"
              cy="17"
              r="2"
            />
          </svg>
          <span
            class="text-sm font-medium"
          >
            4
          </span>
        </div>
        <span
          class="text-xs font-medium"
        >
          Parking
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
